nodejs-week2/app.js

```diff
@@ -0,0 +1,120 @@
+import express from "express";
+import fs from "fs/promises";
+import path from "path";
+import { fileURLToPath } from "url";
+
+const app = express();
+const port = process.env.PORT || 3003;
+const __filename = fileURLToPath(import.meta.url);
+const __dirname = path.dirname(__filename);
+const filePath = path.join(__dirname, "documents.json");
+
+app.use(express.json());
+
+// Middleware for error handling
+const errorHandler = (err, req, res, next) => {
+  console.error(err.stack);
+  res.status(err.status || 500).json({
+    message: err.message || "Internal Server Error",
+  });
+};
+
+// Middleware for loading documents
+let cachedDocuments = null;
+const loadDocuments = async () => {
+  if (!cachedDocuments) {
+    try {
+      const data = await fs.readFile(filePath, "utf-8");
+      cachedDocuments = JSON.parse(data);
+    } catch (err) {
+      console.error("Error parsing JSON file:", err.message);
+      throw new Error("Failed to load documents.");
+    }
+  }
+  return cachedDocuments;
+};
+
+// GET /search
+app.get("/search", async (req, res, next) => {
+  try {
+    const documents = await loadDocuments();
+    const query = req.query.q ? req.query.q.toLowerCase() : null;
+
+    const filteredDocuments = query
+      ? documents.filter(document =>
+          Object.values(document)
+            .map(value => String(value).toLowerCase())
+            .some(value => value.includes(query))
+        )
+      : documents;
```

Comment:
This is quite a lot going on in a single nested command.
You think you can break it apart? in a `if` `else` statement?

Also instead of using `Object.values(document)` is could be a good idea to explicitly choose the fields which should be searched. Then you are also making sure that you dont have to cast to `String`